### PR TITLE
 CI/DOC: Building the documentation with azure-pipelines

### PR DESCRIPTION
**Pull request by datapythonista**
- [X] xref #22766, #26574
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Looks like the problem in #26591 was that the `HostName` input of the azure task that installs the key, is not the host name, but the `know_hosts` line for the host name where to push.

I set up azure-pipelines in my pandas branch, to be able to see the `master` builds and avoid surprises, and this branch is successfully pushing the docs to https://dev.pandas.io

@TomAugspurger since there were differences between the travis docs dependencies and `environment.yml` I think it makes more sense to change and use `environment.yml` here, so we can detect and fix any problem while we still have the travis docs, and then simply remove the travis docs dependencies with the build when we are happy with this.

@jorisvandenbossche the `ImportError` for `pyarrow` doesn't seem to be because `pyarrow` is not in `environment.yml`, it is. There is something weird with that, need to have a look.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
> @jorisvandenbossche the ImportError for pyarrow doesn't seem to be because pyarrow is not in environment.yml, it is. There is something weird with that, need to have a look.

Ah, yes, indeed. Looking at the error message now, it is indeed a problem with the installation, not that it is not installed. 
Possible reason is a mixture of defaults and conda-forge (eg arrow-cpp is from conda-forge, but pyarrow from defaults)

**Comment by datapythonista** [comment]
I think the problem is that `pyarrow` manages to install without a dependency or with something broken, but then it fails on import. We should probably append the original error message to the error we raise.

```
ImportError: /home/vsts/miniconda3/envs/pandas-dev/lib/python3.7/site-packages/pyarrow/lib.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZNK5arrow6Status8ToStringB5cxx11Ev
```

Any idea on what is missing?

**Comment by jorisvandenbossche** [comment]
I think it is related to the mentioned defaults/conda-forge mixing, but why conda is doing that, not sure ..

**Comment by datapythonista** [comment]
I see, not sure why `arrow-cpp` is installed from conda-forge with `environment.yml`. I'm in a windows machine right now, and I get the same, but `pyarrow` is working. Another difference is that with `environment.yml` is using Python 3.7 and not 3.6. But again, in my local set up, `pyarrow` still works with it.

I'll make azure build with `travis-36-doc.yaml` and see if that fixes the problem, otherwise I'm not sure what else we can do.

**Comment by datapythonista** [comment]
Looks like building with `ci/deps/travis-36-doc.yaml` doesn't cause the `pyarrow` problems in azure. Using that here, so we can merge this, and start to build the docs from azure.

I added the missing dependencies in `environment.yml` in #26657 and I opened #26656 for the pyarrow error.

**Comment by datapythonista** [comment]
It's very weird, but the CI was green, and eventually it became red because instead of referencing its own build, was referencing a build of a different PR (https://travis-ci.org/pandas-dev/pandas/jobs/541912722).

I rebased, hope this fix it, but mentioning here in case someone else sees the same, we should probably report the problem to github.

**Comment by jorisvandenbossche** [comment]
The clipboard is not yet working, see here how I recently fixed it on Travis (https://github.com/pandas-dev/pandas/pull/26103/files), but no idea if something similar can fix it on appveyor (might actually also be a good question if the tests for this are working).

**Comment by datapythonista** [comment]
Thanks @jorisvandenbossche I saw the warning, but thought it was one of the warnings that also happens in Travis.

I opened #26700 to address it separately. No change should be required here.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -116,3 +116,65 @@ jobs:
       fi
     displayName: 'Running benchmarks'
     condition: true
+
+- job: 'Docs'
```
Comment (jreback) [inline azure-pipelines.yml]: can you make it so if this build fails it won't fail the entire build? just so we don't replicate the same problem as last time

Comment (datapythonista) [reply azure-pipelines.yml]: Doesn't seem like pipelines has a feature like the travis one for allowed failures.

I can make the `git push` step finish green even if it fails, I guess that's the closer option. But this time I set up the CI in my pandas fork, and made sure the part only executed after merging to master is able to push to dev.pandas.io, and it's working correctly.

Let me know if you want to be extra sure, but shouldn't happen the same as last time, the whole job is tested now.

Comment (jreback) [reply azure-pipelines.yml]: right that's the part that could fail (as its only done on master); maybe not fail even if that fails

**Comment by datapythonista** [comment]
@jreback made the changes so the CI is still green even if the push of the documentation from azure fails. Tested both that the push works, and that if it fails the CI stays green.

Once merged, this will start updating dev.pandas.io with the master version of the docs.

**Comment by datapythonista** [comment]
@jorisvandenbossche can you have a look here? I think would be good to get this merged some time before we activate the fail on warning in the docs.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -116,3 +116,66 @@ jobs:
...
+  pool:
+    vmImage: ubuntu-16.04
+  timeoutInMinutes: 90
+  steps:
+  - script: |
+      echo '##vso[task.setvariable variable=CONDA_ENV]pandas-dev'
```
Comment (jorisvandenbossche) [inline azure-pipelines.yml]: what is this used for? (I don't find it anywhere used in our CI config)

File: azure-pipelines.yml [hunk 3]
```diff
@@ -116,3 +116,66 @@ jobs:
...
+      sshPublicKey: 'ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDHmz3l/EdqrgNxEUKkwDUuUcLv91unig03pYFGO/DMIgCmPdMG96zAgfnESd837Rm0wSSqylwSzkRJt5MV/TpFlcVifDLDQmUhqCeO8Z6dLl/oe35UKmyYICVwcvQTAaHNnYRpKC5IUlTh0JEtw9fGlnp1Ta7U1ENBLbKdpywczElhZu+hOQ892zqOj3CwA+U2329/d6cd7YnqIKoFN9DWT3kS5K6JE4IoBfQEVekIOs23bKjNLvPoOmi6CroAhu/K8j+NCWQjge5eJf2x/yTnIIP1PlEcXoHIr8io517posIx3TBup+CN8bNS1PpDW3jyD3ttl1uoBudjOQrobNnJeR6Rn67DRkG6IhSwr3BWj8alwUG5mTdZzwV5Pa9KZFdIiqX7NoDGg+itsR39QCn0thK8lGRNSR8KrWC1PSjecwelKBO7uQ7rnk/rkrZdBWR4oEA8YgNH8tirUw5WfOr5a0AIaJicKxGKNdMxZt+zmC+bS7F4YCOGIm9KHa43RrKhoGRhRf9fHHHKUPwFGqtWG4ykcUgoamDOURJyepesBAO3FiRE9rLU6ILbB3yEqqoekborHmAJD5vf7PWItW3Q/YQKuk3kkqRcKnexPyzyyq5lUgTi8CxxZdaASIOu294wjBhhdyHlXEkVTNJ9JKkj/obF+XiIIp0cBDsOXY9hDQ== pandas-dev@python.org'
+      sshKeySecureFile: 'pandas_docs_key'
+    displayName: 'Install GitHub ssh deployment key'
+    condition : |
+      and(not(eq(variables['Build.Reason'], 'PullRequest')),
+          eq(variables['Build.SourceBranch'], 'refs/heads/master'))
+
+  - script: |
+      cd doc/build/html
+      git remote add origin git@github.com:pandas-dev/pandas-dev.github.io.git
```
Comment (jorisvandenbossche) [inline azure-pipelines.yml]: At some point you also had a pandas-dev-docs, that's removed again?

**Review by jorisvandenbossche — COMMENTED** [review]
tiny question

File: azure-pipelines.yml (hunk 2 above)
Comment (datapythonista) [reply azure-pipelines.yml]: This is used in `ci/setup_env.sh` to create the conda environment. It's standard in all builds.

File: azure-pipelines.yml (hunk 3 above)
Comment (datapythonista) [reply azure-pipelines.yml]: Yes, I started using that, but then I removed it. Using `<organization-name>.github.io` we're able to use directly a custom domain `dev.pandas.io`. Any other name of repo will make the url something like `dev.pandas.io/pandas-dev-docs`. Personally I think it'll be very nice to have `dev.pandas.io`, even more when we move to `pandas.io`. After almost two years working in the docs I still don't remember the url of the travis docs. :)

File: azure-pipelines.yml (hunk 2 above)
Comment (jorisvandenbossche) [reply azure-pipelines.yml]: I don't see any "CONDA_ENV" in https://github.com/pandas-dev/pandas/blob/master/ci/setup_env.sh, it seems this name is hardcoded everywhre (also in this yaml file a few lines below)

File: azure-pipelines.yml (hunk 3 above)
Comment (jorisvandenbossche) [reply azure-pipelines.yml]: Okido, I am not fully sure about the final url, but that's a discussion for later :-)

File: azure-pipelines.yml (hunk 2 above)
Comment (datapythonista) [reply azure-pipelines.yml]: I think you're right, I think it was removed when `ci/setup.sh` was refactored to be used everywhere and is not needed anymore.

**Comment by datapythonista** [comment]
@jorisvandenbossche addressed your comments, CI is green

**Comment by jorisvandenbossche** [comment]
OK, let's try again! :)

**Comment by datapythonista** [comment]
thanks @jorisvandenbossche, CI is green, and docs where successfully built and pushed to http://dev.pandas.io/ :)

**Pull request merged**